src/Form/Callbacks/PathOrUrlCallbacks.php

```diff
@@ -0,0 +1,22 @@
+<?php
+declare(strict_types = 1);
+
+namespace Drupal\civicrm_newsletter\Form\Callbacks;
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Render\Element\PathElement;
+use Drupal\Core\Render\Element\Url;
+
+final class PathOrUrlCallbacks {
+
+  public static function validate(array &$element, FormStateInterface $formState, array &$form): void {
+    if (is_string($element['#value'] ?? NULL) && preg_match('^[[:alnum:]]+://', ltrim($element['#value']))) {
```

Comment:
Doesn't this interpret protocol-relative URLs like `//example.org/foo/bar` as internal paths?

Reply:
What about checking with `\Drupal\Core\Path\PathValidator::getUrlIfValidWithoutAccessCheck()` which should distinguish between internal and external paths already via `\Drupal\Core\Path\PathValidator::getUrl()`?

Reply:
`PathValidator` which is used to create the `URL`objects before redirect strips leading `/`. So we cannot allow protocol-relative URLs as external URL. (Actually I cannot see a use case for external URLs without scheme.)

By using `Url::validateUrl()` and `PathElement::validateMatchedPath()` we do not need to provide fitting validation messages.

Reply:
Makes sense, I'm just not sure about the extra regex here - do you think this could use `\Drupal\Component\Utility\UrlHelper::isExternal()` instead?

Reply:
Yes, this would be possible. The only difference that I can see is that the validation error message might be different.
E.g. `The URL foo://example.org/ is not valid.` vs
`This path does not exist or you do not have permission to link to foo://example.org/.`

Reply:
The regex is replaced by `\Drupal\Component\Utility\UrlHelper::isExternal()`.